File: frontend/src/services/semesterUserBookService.js
import api from './apiService';

class SemesterUserBookService {
  async getAllSemesterUserBooks() {
    const response = await api.get('/admin/semester-user-books');
    return response.data;
  }

  async getSemesterUserBookById(id) {
    const response = await api.get(`/admin/semester-user-books/${id}`);
    return response.data;
  }

  async getSemesterUserBooksBySemester(semesterId) {
    const response = await api.get(`/admin/semester-user-books/semester/${semesterId}`);
    return response.data;
  }

  async getSemesterUserBooksByPerson(personId) {
    const response = await api.get(`/admin/semester-user-books/person/${personId}`);
    return response.data;
  }

  async getSemesterUserBooksByBook(bookId) {
    const response = await api.get(`/admin/semester-user-books/book/${bookId}`);
    return response.data;
  }

  async getSemesterUserBooksBySemesterAndPerson(semesterId, personId) {
    const response = await api.get(`/admin/semester-user-books/semester/${semesterId}/person/${personId}`);
    return response.data;
  }

  async getSemesterUserBooksBySemesterAndBook(semesterId, bookId) {
    const response = await api.get(`/admin/semester-user-books/semester/${semesterId}/book/${bookId}`);
    return response.data;
  }

  async createSemesterUserBook(semesterId, personId, bookId, data) {
    const response = await api.post(
      `/admin/semester-user-books?semesterId=${semesterId}&personId=${personId}&bookId=${bookId}`,
      data
    );
    return response.data;
  }

  async updateSemesterUserBook(id, data) {
    const response = await api.put(`/admin/semester-user-books/${id}`, data);
    return response.data;
  }

  async deleteSemesterUserBook(id) {
    const response = await api.delete(`/admin/semester-user-books/${id}`);
    return response.data;
  }
}

export default new SemesterUserBookService();

File: src/main/java/com/example/sample/service/SemesterUserBookService.java
package com.example.sample.service;

import com.example.sample.model.Book;
import com.example.sample.model.Person;
import com.example.sample.model.Semester;
import com.example.sample.model.SemesterUserBook;
import com.example.sample.repository.BookRepository;
import com.example.sample.repository.PersonRepository;
import com.example.sample.repository.SemesterRepository;
import com.example.sample.repository.SemesterUserBookRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class SemesterUserBookService {

    private final SemesterUserBookRepository semesterUserBookRepository;
    private final SemesterRepository semesterRepository;
    private final PersonRepository personRepository;
    private final BookRepository bookRepository;

    @Autowired
    public SemesterUserBookService(SemesterUserBookRepository semesterUserBookRepository,
                                   SemesterRepository semesterRepository,
                                   PersonRepository personRepository,
                                   BookRepository bookRepository) {
        this.semesterUserBookRepository = semesterUserBookRepository;
        this.semesterRepository = semesterRepository;
        this.personRepository = personRepository;
        this.bookRepository = bookRepository;
    }

    public List<SemesterUserBook> getAllSemesterUserBooks() {
        return semesterUserBookRepository.findAll();
    }

    public Optional<SemesterUserBook> getSemesterUserBookById(Integer id) {
        return semesterUserBookRepository.findById(id);
    }

    public List<SemesterUserBook> getSemesterUserBooksBySemester(Integer semesterId) {
        return semesterUserBookRepository.findBySemester_SemesterId(semesterId);
    }

    public List<SemesterUserBook> getSemesterUserBooksByPerson(Integer personId) {
        return semesterUserBookRepository.findByPerson_PersonId(personId);
    }

    public List<SemesterUserBook> getSemesterUserBooksByBook(Integer bookId) {
        return semesterUserBookRepository.findByBook_BookId(bookId);
    }

    public List<SemesterUserBook> getSemesterUserBooksBySemesterAndPerson(Integer semesterId, Integer personId) {
        return semesterUserBookRepository.findBySemester_SemesterIdAndPerson_PersonId(semesterId, personId);
    }

    public List<SemesterUserBook> getSemesterUserBooksBySemesterAndBook(Integer semesterId, Integer bookId) {
        return semesterUserBookRepository.findBySemester_SemesterIdAndBook_BookId(semesterId, bookId);
    }

    public SemesterUserBook createSemesterUserBook(Integer semesterId, Integer personId, Integer bookId, SemesterUserBook semesterUserBookData) {
        Semester semester = semesterRepository.findById(semesterId)
                .orElseThrow(() -> new IllegalArgumentException("Semester not found with id: " + semesterId));

        Person person = personRepository.findById(personId)
                .orElseThrow(() -> new IllegalArgumentException("Person not found with id: " + personId));

        Book book = bookRepository.findById(bookId)
                .orElseThrow(() -> new IllegalArgumentException("Book not found with id: " + bookId));

        SemesterUserBook semesterUserBook = new SemesterUserBook();
        semesterUserBook.setSemester(semester);
        semesterUserBook.setPerson(person);
        semesterUserBook.setBook(book);
        semesterUserBook.setStatus(semesterUserBookData.getStatus());
        semesterUserBook.setDate(semesterUserBookData.getDate());

        return semesterUserBookRepository.save(semesterUserBook);
    }

    public SemesterUserBook updateSemesterUserBook(Integer id, SemesterUserBook semesterUserBookData) {
        SemesterUserBook semesterUserBook = semesterUserBookRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("SemesterUserBook not found with id: " + id));

        if (semesterUserBookData.getStatus() != null) {
            semesterUserBook.setStatus(semesterUserBookData.getStatus());
        }

        if (semesterUserBookData.getDate() != null) {
            semesterUserBook.setDate(semesterUserBookData.getDate());
        }

        return semesterUserBookRepository.save(semesterUserBook);
    }

    public void deleteSemesterUserBook(Integer id) {
        if (!semesterUserBookRepository.existsById(id)) {
            throw new IllegalArgumentException("SemesterUserBook not found with id: " + id);
        }
        semesterUserBookRepository.deleteById(id);
    }
}